### PR TITLE
TST: tests for needs-test issues

### LINKED ISSUE #5798: regex filter on non-string columns raises an exception
I'm not really sure what should be done here. At least a better error message I suppose. Perhaps just skip non-string columns or maybe even try to turn them into strings? E.g., make match("\d", 123) work? Obviously could run into some problems for things that can't be converted (or do "unexpected" things) for turning into string/unicode.

``` python
>>> pd.version.version
>>> '0.12.0-1149-g141e93a'

(if columns are all numerical also bombs with different msg)
>>> pd.DataFrame(np.random.random((3,2)), columns=['STRING', 123]).filter(regex='STRING')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-427-e5acd93a79d6> in <module>()
----> 1 pd.DataFrame(np.random.random((3,2)), columns=['STRING', 123]).filter(regex='STRING')

/usr/local/lib/python2.7/dist-packages/pandas-0.12.0_1149_g141e93a-py2.7-linux-x86_64.egg/pandas/core/generic.pyc in filter(self, items, like, regex, axis)
   1522             matcher = re.compile(regex)
   1523             return self.select(lambda x: matcher.search(x) is not None,
-> 1524                                axis=axis_name)
   1525         else:
   1526             raise TypeError('Must pass either `items`, `like`, or `regex`')

/usr/local/lib/python2.7/dist-packages/pandas-0.12.0_1149_g141e93a-py2.7-linux-x86_64.egg/pandas/core/generic.pyc in select(self, crit, axis)
   1116         if len(axis_values) > 0:
   1117             new_axis = axis_values[
-> 1118                 np.asarray([bool(crit(label)) for label in axis_values])]
   1119         else:
   1120             new_axis = axis_values

/usr/local/lib/python2.7/dist-packages/pandas-0.12.0_1149_g141e93a-py2.7-linux-x86_64.egg/pandas/core/generic.pyc in <lambda>(x)
   1521         elif regex:
   1522             matcher = re.compile(regex)
-> 1523             return self.select(lambda x: matcher.search(x) is not None,
   1524                                axis=axis_name)
   1525         else:

TypeError: expected string or buffer
```

### LINKED ISSUE #6795: BUG: rolling_max with center=True and min_periods=1 returns incorrect results
rolling_max with center=True and min_periods=1 returns NaN's 

```
Python 3.3.5 (default, Mar 10 2014, 03:21:31) 
[GCC 4.8.2 20140206 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> df=pd.DataFrame([0,1,2,1,0], columns=['a']

### This is OK
>>> pd.rolling_max(df['a'], window=3, center=True)
0   NaN
1     2
2     2
3     2
4   NaN
dtype: float64


### This should return [1, 2, 2, 2, 1]
>>> pd.rolling_max(df['a'], window=3, center=True, min_periods=1)
0     1
1     2
2     2
3     2
4   NaN
dtype: float64

>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.3.5.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.6-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.13.1
Cython: 0.20.1
numpy: 1.8.1
scipy: 0.13.3
statsmodels: 0.5.0
IPython: 1.2.1
sphinx: 1.2.2
patsy: None
scikits.timeseries: None
dateutil: 2.2
pytz: 2014.1
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.3.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
sqlalchemy: None
lxml: None
bs4: None
html5lib: None
bq: None
apiclient: None

```

### LINKED ISSUE #9752: BUG: first_valid_index() empty Series and DataFrame inconsistencies
I came across some inconsistencies in how `first_valid_index()` works on empty `Series` and `DataFrames` (and `last_valid_index()`).

``` python
>>> print(pd.DataFrame().first_valid_index())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/will/.virtualenvs/dev3/lib/python3.4/site-packages/pandas/core/frame.py", line 3332, in first_valid_index
    return self.index[self.count(1) > 0][0]
  File "/home/will/.virtualenvs/dev3/lib/python3.4/site-packages/pandas/core/index.py", line 915, in __getitem__
    return getitem(key)
IndexError: index 0 is out of bounds for axis 0 with size 0

>>> print(pd.Series().first_valid_index())
None

>>> print(pd.SparseDataFrame([]).first_valid_index())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/will/.virtualenvs/dev3/lib/python3.4/site-packages/pandas/core/frame.py", line 3332, in first_valid_index
    return self.index[self.count(1) > 0][0]
  File "/home/will/.virtualenvs/dev3/lib/python3.4/site-packages/pandas/core/ops.py", line 919, in f
    res = self._combine_const(other, func, raise_on_error=False)
TypeError: _combine_const() got an unexpected keyword argument 'raise_on_error'

>>> print(pd.SparseSeries([]).first_valid_index())
None
```

Is there a reason for this different behavior? Note also that the `DataFrame`/`SparseDataFrame` exceptions are different. Here's the [`DataFrame`](https://github.com/pydata/pandas/blob/671c4b3c52a02ee3736752b2645d08762c0fedaf/pandas/core/frame.py#L3421) and [`Series`](https://github.com/pydata/pandas/blob/a4772028fef7f660cf6d94129a1a9a6f3852e441/pandas/core/series.py#L2375) methods. The `Series` method has a `if len(self) == 0` check. I can make a PR if you think one of these should be changed and tell me which way you want to go. 

For what it's worth, I came across this issue trying to use this function in a list comprehension to get a list of minimum times from a handful of `Series, DataFrames, and SparseDataFrames` that may or may not be empty. (Although it just occurred to me that I could, of course, put an if empty check in my list comprehension.)

### LINKED ISSUE #11220: BUG: numpy.datetime('NaT') printing inconsistency
Whether `np.datetime64('NaT')` is printed as `NaT` or `NaN` depends on other types present in the column. This is consistent with `pd.NaT`, whose display also depends on other types.  But when all the other scalars are `None`, `pd.NaT`  and `None` are both printed as `NaT`, whereas in a column of `np.datetime64('NaT')` and `None`, it depends on the order of the entries (last two lines).  

```
In [44]: pd.DataFrame([[np.datetime64('NaT')], [None]])
Out[44]:
      0
0   NaN
1  None

In [45]: pd.DataFrame([[pd.NaT], [None]])
Out[45]:
    0
0 NaT
1 NaT

In [46]: pd.DataFrame([[np.datetime64('NaT')], [pd.NaT]])
Out[46]:
    0
0 NaT
1 NaT

In [47]: pd.DataFrame([[pd.NaT], [np.datetime64('NaT')]])
Out[47]:
    0
0 NaT
1 NaT

In [81]: pd.DataFrame([[None],[np.datetime64('NaT')]])
Out[81]:
    0
0 NaT
1 NaT

In [82]: pd.DataFrame([[np.datetime64('NaT')],[None]])
Out[82]:
      0
0   NaN
1  None

In [86]: pd.DataFrame([[None],[np.datetime64('NaT')]]).dtypes
Out[86]:
0    datetime64[ns]
dtype: object

In [87]: pd.DataFrame([[np.datetime64('NaT')],[None]]).dtypes
Out[87]:
0    object
dtype: object
```

### LINKED ISSUE #13000: printing multi-index dataframe with renamed index produces traceback
When generating a multi-indexed dataframe, if the default index is renamed and the dataframe is longer than the `max_rows` display option. This cropped up after updating from 0.17 to 0.18. 
#### Code Sample, a copy-pastable example if possible

```
>>> df = pd.DataFrame({'Vals':range(100)})
>>> df = pd.concat([df, ], keys=['Sweep'], names=['Sweep', 'Index'])
>>> df
Traceback (most recent call last):
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\range.py", line 124, in _simple_new
    return RangeIndex(start, stop, step, name=name, **kwargs)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\range.py", line 73, in __new__
    start = _ensure_int(start, 'start')
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\range.py", line 59, in _ensure_int
    new_value = int(value)
TypeError: only length-1 arrays can be converted to Python scalars

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\base.py", line 63, in __repr__
    return str(self)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\base.py", line 42, in __str__
    return self.__unicode__()
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\frame.py", line 534, in __unicode__
    line_width=width, show_dimensions=show_dimensions)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\frame.py", line 1486, in to_string
    show_dimensions=show_dimensions)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\format.py", line 378, in __init__
    self._chk_truncate()
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\core\format.py", line 450, in _chk_truncate
    frame.iloc[-row_num:, :]))
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\tools\merge.py", line 834, in concat
    copy=copy)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\tools\merge.py", line 972, in __init__
    self.new_axes = self._get_new_axes()
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\tools\merge.py", line 1059, in _get_new_axes
    new_axes[self.axis] = self._get_concat_axis()
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\tools\merge.py", line 1111, in _get_concat_axis
    concat_axis = _concat_indexes(indexes)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\tools\merge.py", line 1129, in _concat_indexes
    return indexes[0].append(indexes[1:])
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\multi.py", line 1029, in append
    label = self.get_level_values(i)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\multi.py", line 666, in get_level_values
    tz=getattr(unique, 'tz', None))
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\range.py", line 126, in _simple_new
    return Index(start, stop, step, name=name, **kwargs)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\base.py", line 214, in __new__
    return Int64Index(data, copy=copy, dtype=dtype, name=name)
  File "C:\Users\User\Anaconda3\lib\site-packages\pandas\indexes\numeric.py", line 107, in __new__
    not issubclass(np.dtype(dtype).type, np.integer)):
TypeError: data type "Index" not understood
```
#### Expected Output

The expected outcome is that the dataframe should print out without producing a traceback, as occurs when the length is less than that set by `max_rows`. I.e.

```
>>> df = pd.DataFrame({'Vals':range(10)})
>>> df = pd.concat([df, ], keys=['Sweep'], names=['Sweep', 'Index'])
>>> df
             Vals
Sweep Index
Sweep 0         0
      1         1
      2         2
      3         3
      4         4
      5         5
      6         6
      7         7
      8         8
      9         9
```

The issue only seems to be with changing the default index name, as you can change the other index name without issue:

```
>>> df = pd.DataFrame({'Vals':range(100)})
>>> df = pd.concat([df, ], keys=['Sweep'], names=['Sweep',])
>>> df
          Vals
Sweep
Sweep 0      0
      1      1
      2      2
      3      3
      4      4
      5      5
      6      6
      7      7
      8      8
      9      9
      10    10
      11    11
      12    12
      13    13
      14    14
      15    15
      16    16
      17    17
      18    18
      19    19
      20    20
      21    21
      22    22
      23    23
      24    24
      25    25
      26    26
      27    27
      28    28
      29    29
...        ...
      70    70
      71    71
      72    72
      73    73
      74    74
      75    75
      76    76
      77    77
      78    78
      79    79
      80    80
      81    81
      82    82
      83    83
      84    84
      85    85
      86    86
      87    87
      88    88
      89    89
      90    90
      91    91
      92    92
      93    93
      94    94
      95    95
      96    96
      97    97
      98    98
      99    99

[100 rows x 1 columns]
```
#### output of `pd.show_versions()`

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.0
setuptools: 20.2.2
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.0.3
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.6
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.5.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.11
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
```

Discussed in more detail [here](http://stackoverflow.com/questions/36869745/pandas-0-18-setting-multi-index-name-on-df-from-read-csv-resulting-in-typee?noredirect=1#comment61309343_36869745)

### LINKED ISSUE #15286: MultiIndex.from_product throws on read-only array
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
a = np.array(range(3))
b = ['a', 'b']
# Succeeds, as expected
pd.MultiIndex.from_product([a, b]) 
a.setflags(write=False)
# Raises a ValueError with "buffer source array is read-only"
pd.MultiIndex.from_product([a, b]) 
```
#### Problem description

Nothing in the MultiIndex.from_product documentation suggests the inputs need to be mutable.  Moreover, it is surprising that the arguments are being mutated.

#### Expected Output

Not raising.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.1.17-pv-ts1
machine: x86_64
processor: 
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2-ts2
nose: 1.3.7
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.11.3
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.4
boto: None
pandas_datareader: None
</details>

### LINKED ISSUE #17198: AttributeError: 'IntBlock'/'FloatBlock'/etc. object has no attribute 'sp_index'
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
# AttributeError: 'IntBlock' object has no attribute 'sp_index'
s = pd.SparseSeries([1, 2])
print(s.where(s >= 2, 0))
# AttributeError: 'FloatBlock' object has no attribute 'sp_index'
df = pd.SparseDataFrame([[1, 2], [3, 4]])
print(df.quantile())
```
#### Problem description

On the surface, SparseSeries and SparseDataFrame should produce the same results as Series and DataFrame, respectively.

#### Expected Output

0    0
1    2
dtype: int64
0    2.0
1    3.0
Name: 0.5, dtype: float64

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 36.0.1
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #5798
- [x] closes #6795
- [x] closes #9752
- [x] closes #11220
- [x] closes #13000
- [x] closes #15286
- [x] closes #17198
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_timeseries.py [hunk 1]
```diff
@@ -609,6 +609,17 @@ def test_first_last_valid(
...
+        # GH#9752 Series/DataFrame should both return None, not raise
+        ser = pd.Series([np.nan])
+        df = pd.DataFrame([np.nan])
+
+        assert ser.first_valid_index() is None
+        assert df.first_valid_index() is None
+
+        assert ser.iloc[:0].first_valid_index() is None
+        assert df.iloc[:0].first_valid_index() is None
+
```
Comment (gfyoung) [inline pandas/tests/frame/test_timeseries.py]: I would parameterize over the object used (i.e. `Series` or `DataFrame`).

File: pandas/tests/indexes/multi/test_constructor.py [hunk 2]
```diff
@@ -577,6 +577,17 @@ def test_from_product_respects_none_names():
...
+    # GH#15286 passing read-only array to from_product
+    a = np.array(range(3))
+    b = ["a", "b"]
+    expected = pd.MultiIndex.from_product([a, b])
+
+    a.setflags(write=False)
+    result = pd.MultiIndex.from_product([a, b])
+    tm.assert_index_equal(result, expected)
+
+
```
Comment (gfyoung) [inline pandas/tests/indexes/multi/test_constructor.py]: `MultiIndex` is imported directly FYI.

File: pandas/tests/io/formats/test_format.py [hunk 3]
```diff
@@ -1106,6 +1106,15 @@ def test_truncate_with_different_dtypes(self):
...
 
+    def test_truncate_with_different_dtypes_multiindex(self):
+        # GH#13000
+        df = pd.DataFrame({"Vals": range(100)})
+        frame = pd.concat([df], keys=["Sweep"], names=["Sweep", "Index"])
+        result = repr(frame)
+
+        result2 = repr(frame.iloc[:5])
+        assert result.startswith(result2)
+
```
Comment (gfyoung) [inline pandas/tests/io/formats/test_format.py]: `DataFrame` is imported directly already FYI.

File: pandas/tests/frame/test_constructors.py [hunk 4]
```diff
@@ -1726,10 +1726,16 @@ def test_constructor_with_datetimes(self):
         tm.assert_frame_equal(df, expected)
 
     def test_constructor_datetimes_with_nulls(self):
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py]: can parameterize (later ok too)

**Review by jreback — COMMENTED** [review]
lgtm. couple of comments, ping on green.

**Comment by jbrockmendel** [comment]
ping

File: pandas/tests/frame/test_constructors.py (hunk 4 above)
Comment (gfyoung) [reply pandas/tests/frame/test_constructors.py]: @jbrockmendel : do you plan on addressing this here or later?

Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py]: later

**Comment by WillAyd** [comment]
Thanks @jbrockmendel great work on this

**Comment by jreback** [comment]
thanks @jbrockmendel if you can parameterize as indicated would be great

**Pull request merged**